xdsl/frontend/pyast/context.py

```diff
@@ -1,12 +1,107 @@
 import ast
+from collections.abc import Callable
 from contextlib import AbstractContextManager
-from dataclasses import dataclass
-from inspect import getsource
+from dataclasses import dataclass, field
+from inspect import currentframe, getsource
 from sys import _getframe  # pyright: ignore[reportPrivateUsage]
-from typing import Any
+from typing import Any, overload
 
-from xdsl.frontend.pyast.program import FrontendProgram
+from xdsl.frontend.builder import PyASTBuilder
+from xdsl.frontend.pyast.program import FrontendProgram, P, PyASTProgram, R
 from xdsl.frontend.pyast.utils.python_code_check import PythonCodeCheck
+from xdsl.frontend.pyast.utils.type_conversion import FunctionRegistry, TypeRegistry
+from xdsl.ir import Operation, TypeAttribute
+
+
+@dataclass
+class PyASTContext:
+    """Encapsulate the mapping between Python and IR types and operations."""
+
+    type_registry: TypeRegistry = field(default_factory=TypeRegistry)
+    """Mappings between source code and IR type."""
+
+    function_registry: FunctionRegistry = field(default_factory=FunctionRegistry)
+    """Mappings between functions and their operation types."""
+
+    def register_type(
+        self,
+        source_type: type,
+        ir_type: TypeAttribute,
+    ) -> None:
+        """Associate a type in the source code with its type in the IR."""
+        self.type_registry.insert(source_type, ir_type)
+
+    def register_function(
+        self, function: Callable[..., Any], ir_constructor: Callable[..., Operation]
+    ) -> None:
+        """Associate a method on an object in the source code with its IR implementation."""
+        self.function_registry.insert(function, ir_constructor)
+
+    @overload
+    def parse_program(
+        self,
+        decorated_func: None = None,
+        *,
+        desymref: bool = True,
+    ) -> Callable[[Callable[P, R]], PyASTProgram[P, R]]: ...
+
+    @overload
+    def parse_program(
+        self,
+        decorated_func: Callable[P, R],
+        *,
+        desymref: bool = True,
+    ) -> PyASTProgram[P, R]: ...
+
+    def parse_program(
+        self,
+        decorated_func: Callable[P, R] | None = None,
+        *,
+        desymref: bool = True,
+    ) -> Callable[[Callable[P, R]], PyASTProgram[P, R]] | PyASTProgram[P, R]:
+        """Get a program wrapper by decorating a function."""
+
+        def decorator(func: Callable[P, R]) -> PyASTProgram[P, R]:
```

Comment:
Would be nice to preserve doc strings etc from the original, let's add a test for the doc string also?
```suggestion
        @functools.wraps(decorated_func)
        def decorator(func: Callable[P, R]) -> PyASTProgram[P, R]:
```
https://docs.python.org/3/library/functools.html#functools.wraps

Reply:
Because the `decorated_func` can be `None` to support invocation as both `ctx.parse_program` and `ctx.parse_program()`, the `@functools.wraps` decorator cannot be used.

I've instead done something which I think should be equivalent directly calling `functools.update_wrapper` within the decorator.

Reply:
Nice, can you please print `test_func.__doc__` to make sure it works?

Reply:
Standalone expressions are not currently supported by the AST parser, so any test cases including docstrings fail, but have tested it manually without the parsing step and the docstrings are preserved, and added `assert program.__doc__ == func.__doc__` which should also guarantee this